File: iridium-angular-client/projects/ngx-iridium-client/src/lib/service/authorization.service.ts
import { Inject, Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';
import { catchError } from 'rxjs/operators';
import { throwError } from 'rxjs';
import { ApiDataResponse } from '../domain/api-data-response';
import { AccessTokenResponse } from '../domain/access-token-response';
import { OauthConstants } from './oauth-constants';
import { CookieService } from './cookie.service';
import { AbstractBaseService } from './abstract-base-service';
import { IdentityResponse } from '../domain/identity-response';

@Injectable({
  providedIn: 'root'
})
export class AuthorizationService extends AbstractBaseService {

  constructor(protected http: HttpClient,
              @Inject('config') private config:any,
  private cookieService: CookieService) {
    super();
  }

  exchangeForExternalIdentity(code: string | null, state: string | null) {
    const headers = new HttpHeaders({
      Accept:  'application/json',
    });

    const httpOptions = {
      headers,
    };
    const redirectUri = this.config.iridium.redirectUri;
    const clientId = this.config.iridium.clientId;

    const codeVerifier = this.cookieService.getCookie(OauthConstants.PKCE_CODE_VERIFIER);
    const url = this.config.iridium.domain
      + 'oauth/token?grant_type=authorization_code&code='
      // tslint:disable-next-line:max-line-length
      + code + '&redirect_uri=' + redirectUri + '&client_id=' + clientId + '&code_verifier=' + codeVerifier + '&state=' + state
    return this.http.post<AccessTokenResponse>(url,null, httpOptions).toPromise();
  }

  exchange(code: string | null) {
    const headers = new HttpHeaders({
      Accept:  'application/json',
    });

    const httpOptions = {
      headers,
    };
    const redirectUri = this.config.iridium.redirectUri;
    const clientId = this.config.iridium.clientId;

    const codeVerifier = this.cookieService.getCookie(OauthConstants.PKCE_CODE_VERIFIER);

    return this.http.post<AccessTokenResponse>(
      this.config.iridium.domain
      + 'oauth/token?grant_type=authorization_code&code='
      + code + '&redirect_uri=' + redirectUri + '&client_id=' + clientId + '&code_verifier=' + codeVerifier, null, httpOptions).toPromise()
  }

  getIdentity(bearerToken: string) {
    const headers = new HttpHeaders({
      Accept:  'application/vnd.iridium.id.identity-response.1+json',
      Authorization: 'Bearer ' + bearerToken
    });

    const httpOptions = {
      headers,
    };

    return this.http.get<IdentityResponse>(
      this.config.iridium.domain
      + 'identities', httpOptions).toPromise()
  }
}

File: iridium-flask-client/service/authorization_service.py
# python class for authorization service     

import os
import json
import http.client

import oauth_constants

class AuthorizationService:
    # python constructor taking http client as parameter
    def __init__(self, http_client):
        # python set http client
        self.http_client = http_client  
        self.base_url = os.environ.get(oauth_constants.PUBLIC_IRIDIUM_DOMAIN)
        self.redirect_uri = os.environ.get(oauth_constants.PUBLIC_IRIDIUM_REDIRECT_URI)) 
        self.client_id = os.environ.get(oauth_constants.PUBLIC_IRIDIUM_CLIENT_ID)


    # python function taking code and state and retrieving external identity
    def get_external_identity(self, code, state):
        # python set headers for http client
        headers = {'Content-type': 'application/json'}
        # python set payload for http client
        payload = {'code': code, 'state': state}

        # get cookie from http_client request
        codeVerifier = self.http_client.get_cookie(oauth_constants.PKCE_CODE_VERIFIER)

        # python set url for http client
        url = (self.base_url + 
               'oauth/token?'+
               'grant_type=authorization_code' + 
               code + '&redirect_uri=' + 
               self.redirect_uri + '&client_id=' + 
               self.client_id + '&code_verifier=' + 
               codeVerifier + '&state=' + state)
        
        # python set method for http client
        method = 'POST'
        # python set body for http client
        body = json.dumps(payload)
        # python set response for http client
        response = self.http_client.request(method, url, body, headers)
        # python return response
        return response

    # python function for key exchange with a code for input 
    def exchange(self, code):
        # python set headers for http client
        headers = {'Content-type': 'application/json'}
        # python set payload for http client
        payload = {'code': code}

        # get cookie from http_client request
        codeVerifier = self.http_client.get_cookie(PKCE_CODE_VERIFIER)

        # python set url for http client
        url = (self.base_url + 
               'oauth/token?'+
               'grant_type=authorization_code' + 
               code + '&redirect_uri=' + 
               self.redirect_uri + '&client_id=' + 
               self.client_id + '&code_verifier=' + 
               codeVerifier)
        
        # python set method for http client
        method = 'POST'
        # python set body for http client
        body = json.dumps(payload)
        # python set response for http client
        response = self.http_client.request(method, url, body, headers)
        # python return response
        return response 

    # python function for taking a bearer token and returning response 
    def get_identity(self, bearer_token):
        # python set headers for http client
        headers = {
            'Accept':  'application/vnd.iridium.id.identity-response.1+json',
            'Authorization': 'Bearer ' + bearer_token
            }
        # python set url for http client
        url = self.base_url + 'identities'
        # python set method for http client
        method = 'GET'
        # python set body for http client
        body = None
        # python set response for http client
        response = self.http_client.request(method, url, body, headers)
        # python return response
        return response